crates/sui-indexer/src/indexer_reader.rs

```diff
@@ -1256,6 +1265,162 @@ impl IndexerReader {
+            let sender = sender_signed_data_option
+                .as_ref()
+                .ok_or_else(|| {
+                    IndexerError::GenericError("Sender signed data should be available".into())
+                })?
+                .intent_message()
+                .value
+                .sender();
+            let effects = effects_option
+                .as_ref()
+                .ok_or_else(|| IndexerError::GenericError("Effects should be available".into()))?;
+            Some(
+                get_object_changes(
+                    self,
+                    sender,
+                    effects.modified_at_versions(),
+                    effects.all_changed_objects(),
+                    effects.all_removed_objects(),
+                )
+                .await?,
+            )
+        } else {
+            None
+        };
+
+        let transaction = if options.show_input {
+            let data = sender_signed_data_option.ok_or_else(|| {
+                IndexerError::GenericError("Sender signed data should be available".into())
+            })?;
+            Some(
+                SuiTransactionBlock::try_from_with_package_resolver(
+                    data.clone(),
+                    package_resolver.clone(),
+                )
+                .await?,
+            )
+        } else {
+            None
+        };
+
+        let effects = if options.show_effects {
+            let effects = effects_option
+                .ok_or_else(|| IndexerError::GenericError("Effects should be available".into()))?;
+            Some(SuiTransactionBlockEffects::try_from(effects)?)
+        } else {
+            None
+        };
+
+        let raw_transaction = if options.show_raw_input {
+            stored_transaction.raw_transaction
+        } else {
+            Vec::new()
+        };
+
+        let events = if options.show_events {
+            let events = {
+                stored_transaction
+                        .events
+                        .into_iter()
+                        .map(|event| match event {
+                            Some(event) => {
+                                let event: Event = bcs::from_bytes(&event).map_err(|e| {
+                                    IndexerError::PersistentStorageDataCorruptionError(format!(
+                                        "Can't convert event bytes into Event. tx_digest={:?} Error: {e}",
+                                        tx_digest
+                                    ))
+                                })?;
+                                Ok(event)
+                            }
+                            None => Err(IndexerError::PersistentStorageDataCorruptionError(format!(
+                                "Event should not be null, tx_digest={:?}",
+                                tx_digest
+                            ))),
+                        })
+                        .collect::<Result<Vec<Event>, IndexerError>>()?
```

Comment:
I wonder if we should instead be storing the whole vector of events serialized together vs having them be a vec of serialized bytes...

Reply:
that sounds very reasonable b/c we do not need the cardinality info. fwiw this part is moved from transaction.rs, but I am happy to do it in a separate pr.

Reply:
actually I was wrong, graphql uses the cardinality info for pagination purpose for example https://github.com/MystenLabs/sui/blob/6bfc91b2c4cda1640fadd7ee71058420d8a63c7b/crates/sui-graphql-rpc/src/types/transaction_block_effects.rs#L362

Reply:
and same thing for events cardinality